### ISSUE TITLE
Scheduler: same Pod can exist in both scheduling cycle and scheduling queue

### ISSUE THREAD

**Issue by sanposhiho** [kubernetes/kubernetes#118226]
### What happened?

1. A Pod named `PodA` is being scheduled.
2. PodA is updated by someone.
3. The scheduler's EventHandler notices the update in PodA. And let the scheduling queue know about the update via `PriorityQueue.Update` method.
4. `PriorityQueue.Update` checks where PodA exists in the scheduling queue. But, PodA is being scheduled and doesn't exist in scheduling queue.
5. `PriorityQueue.Update` adds the PodA to activeQ. ([ref](https://github.com/kubernetes/kubernetes/blob/0bff705acd8982e34b937116eb2016c9d6e4c4a6/pkg/scheduler/internal/queue/scheduling_queue.go#L680-L684))
6. PodA is being scheduled but exists in the activeQ at the same time.

---

With a similar scenario, it could also happen that the same Pod can exist in both the binding cycle and the scheduling queue, or in both the binding cycle and the scheduling cycle.

### What did you expect to happen?

If the Pod is being scheduled, the scheduling queue shouldn't append the Pod to activeQ.
This bug is easy to solve after in-flight Pods support in https://github.com/kubernetes/kubernetes/pull/117844.

### How can we reproduce it (as minimally and precisely as possible)?

Update a Pod during a Pod is being scheduled.

### Anything else we need to know?

/assign
/sig scheduling

### Kubernetes version

master

### Cloud provider

<details>

</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by NoicFank** [comment]
Do you mean assumed pod？

It seems that it has already considered.
https://github.com/kubernetes/kubernetes/blob/b374404825125f5ea8c46313ccaf3717e6ba46fb/pkg/scheduler/schedule_one.go#L309-L327

**Comment by sanposhiho** [comment]
No, literally two Pods (one's in the scheduling cycle, another's put in the scheduling queue.) can exist. 
See:
https://github.com/kubernetes/kubernetes/blob/0bff705acd8982e34b937116eb2016c9d6e4c4a6/pkg/scheduler/internal/queue/scheduling_queue.go#L680-L684

**Comment by NoicFank** [comment]
Okay, I got it. 

Will this case have any negative impact?

**Comment by sanposhiho** [comment]
> literally two Pods (one's in the scheduling cycle, another's put in the scheduling queue.) can exist. 

I expect there are only minor negative impacts.

_For ease of explanation, name "one's in the scheduling cycle" Pod1, name "another's put in the scheduling queue" Pod2._

If Pod1 is successfully scheduled:
- weird `FailedScheduling` message: Pod2 will proceed to the scheduling cycle, but fail in `assume`. ([ref](https://github.com/sanposhiho/kubernetes/blob/8badc6f8fe45e2133ca23ba00ecba94dbe04c8d8/pkg/scheduler/internal/cache/cache.go#L381))

If Pod1 is failed to get scheduled:
- We enqueue Pod back to activeQ only when the Pod likely get scheduled by the event. So, the scheduling of Pod2 is useless and very likely rejected for the same reason as Pod1. Such useless retry will increase the backoff time.
- [only when the scheduling of Pod2 has actually failed] the Pod in the scheduling queue is old during short time because Pod2 will be pushed back to the scheduling queue, but it's failed ([ref](https://github.com/sanposhiho/kubernetes/blob/8badc6f8fe45e2133ca23ba00ecba94dbe04c8d8/pkg/scheduler/internal/queue/scheduling_queue.go#L601-L610)). Though this old Pod will be updated soon by Pod updated event from [here](https://github.com/sanposhiho/kubernetes/blob/8badc6f8fe45e2133ca23ba00ecba94dbe04c8d8/pkg/scheduler/schedule_one.go#L940).

**Comment by NoicFank** [comment]
Yes, I do a sample test for case 1, that if pod1 is successfully `assigned` in scheduler cache (already propagate back from apiServer, before that, pod1 will remains `assumed`, and assumed is fine). Then pod2 joins the scheduling cycle will trigger the following error:
```
Scheduler cache AssumePod failed" err="pod scheduled-pod1(scheduled-pod1) is in the cache, so can't be assumed
```

And, I think this could be improved within func, that bypass those already assigned pod.
https://github.com/sanposhiho/kubernetes/blob/8badc6f8fe45e2133ca23ba00ecba94dbe04c8d8/pkg/scheduler/schedule_one.go#L328

**Comment by sanposhiho** [comment]
The root cause of this issue is that the scheduling queue doesn't know which Pod is being scheduled now and put the Pod during scheduling into the activeQ.
So, as I said at the top, we should wait for the in-flight Pod support, which enable the queue to know which Pod is being scheduled. After that, we can easily ignore the update of Pod during scheduling in the queue.

**Comment by wackxu** [comment]
I want to help fix it
/assign

**Comment by kerthcet** [comment]
We already supported inflight pods, is this still a concern? @sanposhiho

**Comment by sanposhiho** [comment]
(I noticed https://github.com/kubernetes/kubernetes/pull/119607 is closed and someone has to take over, but https://github.com/kubernetes/kubernetes/pull/122234 has to be merged first, so I just leave it for now)

**Comment by nayihz** [comment]
I see #119607 has been closed but no one take over it. Pls assign to me if nobody create new pr for this issue. @sanposhiho

**Comment by sanposhiho** [comment]
Yes, please take over.

/assign @nayihz

**Comment by nayihz** [comment]
Open https://github.com/kubernetes/kubernetes/pull/125578 for this issue. PTAL if you have time. Thanks.

**Comment by nayihz** [comment]
I think we can close it after #125578 merged.
/close
